### PR TITLE
Fix problems with assoc expr token collection

### PR DESCRIPTION
**Pull request by nnethercote**
There are several cases involving assoc exprs and attributes where the current code does the wrong thing. This PR adds some tests that demonstrate the problems and then fixes them.

r? @petrochenkov

### CODE REVIEW

**Comment by programmerjake** [comment]
would you be willing to add some tests that you can invoke attribute proc macros where the input has attributes on expressions without needing nightly features? (I wasn't able to find a test for that, and that's already stable behavior that the library I'm writing depends on).

e.g. something like
```rust
// no nightly features needed
#[check_input_tokens]
fn f() {
    #[a] a + #[b] b + #[c] c;
}
```

File: compiler/rustc_parse/src/parser/expr.rs [hunk 1]
```diff
@@ -59,15 +59,30 @@ impl<'a> Parser<'a> {
...
         let attrs = self.parse_outer_attributes()?;
-        self.parse_expr_res(Restrictions::empty(), attrs)
+        self.parse_expr_res(Restrictions::empty(), attrs).map(|res| res.0)
     }
 
     /// Parses an expression, forcing tokens to be collected.
     pub fn parse_expr_force_collect(&mut self) -> PResult<'a, P<Expr>> {
         self.current_closure.take();
 
+        // If the expression is associative (e.g. `1 + 2`), then any preceding
```
Comment (petrochenkov) [inline compiler/rustc_parse/src/parser/expr.rs:69]: Attributes have the priority of an unary operator (like `-`, `!`, `*`, `&`, `box`), or at least they are supposed to.
So these interactions may affect not only binary operators (including assignments), but also ranges `#[attr] a .. b` (binds to `a`), casts `#[attr] a as Type` (binds to `a`), maybe something else.

I'm not sure, is all that covered by "expression is associative"?

**Comment by nnethercote** [comment]
Performance effects are negligible, which matches what I saw locally. This is good because this change does increase the amount of `TokenCursor` snapshot-taking.

File: compiler/rustc_parse/src/parser/mod.rs [hunk 2]
```diff
@@ -237,6 +237,7 @@ impl NodeRange {
     // is the position of the function's start token. This gives
     // `NodeRange(10..15)`.
     fn new(ParserRange(parser_range): ParserRange, start_pos: u32) -> NodeRange {
+        assert!(parser_range.start >= start_pos);
```
[the remark this answers is not in the document]
Comment (nnethercote) [reply compiler/rustc_parse/src/parser/mod.rs:240]: Ok.

File: compiler/rustc_parse/src/parser/expr.rs [hunk 3]
```diff
@@ -255,6 +274,7 @@ impl<'a> Parser<'a> {
                 break;
             }
 
+            parsed_something = true;
```
[the remark this answers is not in the document]
Comment (nnethercote) [reply compiler/rustc_parse/src/parser/expr.rs:277]: Good catch. I will move this assignment higher up.

File: compiler/rustc_parse/src/parser/expr.rs (hunk 1 above)
Comment (nnethercote) [reply compiler/rustc_parse/src/parser/expr.rs:69]: I think so. `AssocOp::from_token` includes `..`, `..=`, `...`, and `as`.

**Comment by nnethercote** [comment]
I have addressed the comments. I also tweaked the `stringify.rs` test to use operators other than just `+`, including `..` and `as`.

**Comment by petrochenkov** [comment]
Let's do a crater run (to check if something like https://github.com/rust-lang/rust/pull/128725#issuecomment-2270923740 depends on the current behavior, in particular).
@bors try

**Comment by petrochenkov** [comment]
@craterbot check

**Comment by petrochenkov** [comment]
No regressions related to this change.
@bors r+

**Comment by nnethercote** [comment]
I rebased.

@bors r=petrochenkov

**Comment by nnethercote** [comment]
I'm not sure if the regressions are real -- a pre-merge CI run showed fewer regressions. And if they are real, they are small and few, and this is fixing a correctness issue.

@rustbot label: +perf-regression-triaged

**Pull request merged**